### ISSUE TITLE
[nn.Module] Non-recursive `_apply()`

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#97049]
**Overview**
This is a feature request for a version of `_apply()` that does not call recursively and instead only applies to the target module, not its submodules. In other words, it skips the first two lines in its current implementation:
https://github.com/pytorch/pytorch/blob/985fc66b308c37bed4c0337e795de4abd929a97a/torch/nn/modules/module.py#L795-L797

For parameter offloading schemes, modules' parameters may no longer have uniform device. Being able to call an API similar to the existing `.to()` but only applying to a target module and not its submodules facilitates clean parameter offloading implementations.

**Case Study: `FullyShardedDataParallel`**
When a user passes both `cpu_offload=CPUOffload(True)` and `device_id` to PyTorch `FullyShardedDataParallel` (FSDP), where there are nested FSDP instances, FSDP calls `module.to(device)`, which incorrectly moves _all_ module's parameters to `device` including submodules whose parameters have already been offloaded to CPU. This greatly reduces the model size that FSDP can support because already sharded parameters in submodules are moved to GPU when they should stay on CPU. The `device_id` argument can help speed up FSDP initialization when CPU offloading is enabled since the flattening and sharding happens on GPU, not CPU.

We want to replace the `module.to(device)` call with a version that traverses `module.modules()` but does not proceed into any existing FSDP module and only moves the traversed modules to `device`. This is where the non-recursive `_apply()` / `to()` helps. Then, we will be able to support much larger models with CPU offloading enabled and `device_id` specified.

**API**
Since `_apply()` is the basis for several public frontends like `.to()`, `.cuda()`, `.cpu()`, adding a `recurse: bool` argument to all may be a sweeping change. Hence, another option is to have a private API like `_apply_non_recursive(fn)` that subsumes the current `_apply()` implementation minus the first two lines:
```
def _apply_non_recursive(fn):
    # Current `_apply()` implementation minus first two lines
    ...

def _apply(fn):
    for module in self.children(): 
        module._apply(fn)
    return _apply_non_recursive(fn)
```

Alternatively, we (the distributed team) could effectively fork the current `_apply()` implementation in distributed code to unblock our FSDP use case. Regardless, any alternative seems to require peaking into the `module._parameters` dict ourselves.

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by awgu** [comment]
> Why is peaking at module._parameters bad? You can access what you need with public APIs, right?

Ah, good point. `get_parameter()` and `register_parameter()` effectively let us look at `module._parameters`.

---

To clarify, I do not see any technical blocker for FSDP to not align with [`compute_should_use_set_data()`](https://github.com/pytorch/pytorch/blob/985fc66b308c37bed4c0337e795de4abd929a97a/torch/nn/modules/module.py#L799) in `_apply()` and instead always set `.data` or always create new parameter variables. However, in case I am missing something, I thought it would be preferred to align with `nn.Module` if possible (meaning set the parameter `.data` iff  `nn.Module._apply()` does). If we want to align with `nn.Module`, then directly using an `nn.Module` API seems better from a maintenance perspective than for us to write our own.

**Comment by awgu** [comment]
For FSDP, we worked around the issue (https://github.com/pytorch/pytorch/pull/98319), and this feature would just be a nice-to-have now.
